boot/freeldr/bootsect/pc98/fat12fdd.S

```diff
@@ -0,0 +1,579 @@
+/*
+ * PROJECT:     FreeLoader
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     FAT12 file system boot sector for NEC PC-98 series
+ * NOTES:       The source code in this file is based on the Brian Palmer's work
+ *              (boot/freeldr/bootsect/fat.S)
+ * COPYRIGHT:   Copyright 2019 Dmitry Borisov (di.sean@protonmail.com)
+ */
+
+#include <asm.inc>
+#include <freeldr/include/arch/pc/x86common.h>
+
+#define BP_REL(x) [bp + x - offset start]
+#define VGA_WIDTH 80
+#define VGA_HEIGHT 25
+
+/*
+ * See https://www.webtech.co.jp/company/doc/undocumented_mem/memsys.txt
+ * Note - DS register must be 0000h
+ */
+#define BOOT_DAUA HEX(0584)
+
+/*
+ * Text video RAM is located at segments 0A000h (characters) and 0A200h (attributes)
+ * See https://people.freebsd.org/~kato/pc98-arch.html#memmap
+ */
+#define VRAM_TEXTMODE_ADDRESS HEX(0A000)
+
+DataAreaStartHigh    = 2
+DataAreaStartLow     = 4
+BiosCHSDriveSizeHigh = 6
+BiosCHSDriveSizeLow  = 8
+BiosCHSDriveSize     = 8
+ReadSectorsOffset    = 10
+ReadClusterOffset    = 12
+PutCharsOffset       = 14
+BootSectorStackTop   = HEX(7C00) - 16
+
+.macro DEBUG_STOP
```

Comment:
Debugging? If so it should go into some #if 0 of sorts.

Reply:
Improved.